[Getting started][Feature flags] Add client side bootstrapping page (#6522)
PostHog/posthog.com: contents/docs/feature-flags/bootstrapping.mdx

---
title: Client-side bootstrapping
sidebar: Docs
showTitle: true
availability:
    free: none
    selfServe: full
    enterprise: full
---

> **Note:** Bootstrapping feature flags is only available in our [JavaScript web](/docs/libraries/js) and [React Native](/docs/libraries/react-native) SDKs.

> Since is a delay between initializing PostHog and fetching feature flags, feature flags are not always available immediately.This can be detrimental if you want to do something like redirecting to a different page based on a feature flag.

Reviewer: ```suggestion
Since there is a delay between initializing PostHog and fetching feature flags, feature flags are not always available immediately. This can be detrimental if you want to do something like redirecting to a different page based on a feature flag.
```

To have your feature flags available immediately, you can initialize PostHog with initial values until PostHog has had a chance to fetch them. This is called bootstrapping.

> To bootstrap PostHog with feature flags, use the `bootstrap` key in the initialization config:

Reviewer 2: ```suggestion
To bootstrap PostHog with feature flags, use the `bootstrap` key in the initialization config and add `featureFlags` values to it:
```

<MultiLanguage>

```js-web
posthog.init('<ph_project_api_key>', {
    api_host: '<ph_instance_address>',
    bootstrap: {
        featureFlags: {
            'flag-1': true,
            'variant-flag': 'control',
            'other-flag': false,
        },
    },
})
```

```react-native
// App.(js|ts)
import { usePostHog, PostHogProvider } from 'posthog-react-native'

export function MyApp() {
    return (
        <PostHogProvider apiKey="<ph_project_api_key>" options={{
            host: '<ph_instance_address>',
            bootstrap: {
                featureFlags: {
                    'flag-1': true,
                    'variant-flag': 'control',
                    'other-flag': false
                }
            }
        }}>
            <MyComponent />
        </PostHogProvider>
    )
}
```

</MultiLanguage>

## Setting the correct flag values

> To ensure you are bootstrapping PostHog with the correct flag values, we recommend fetching the flags values on your server and then passing them to your frontend.

Reviewer: ```suggestion
To ensure you are bootstrapping PostHog with the correct flag values, we recommend fetching the flags values from your server alongside the page load request, and then passing them to your frontend.
```

It wouldn't make sense if this is an additional request, as the delay still exists then

You can do this by:

> 1. On your frontend, make a request to your server
> 2. Your server fetches all the feature flags for the user by calling `getAllFlags()`
> 3. Your server returns the flag values in the response

Reviewer 2: This is a bit confusing to explain, but basically "frontend" request go to a server then return the frontend. Bootstrapping flag request should be handled along side the frontend request. 

```suggestion
1. When receiving a frontend request, fetch all the feature flags for the user by calling `getAllFlags()`.
2. Return the frontend request with the flags values in the response.
3. On the frontend, get the flag values from the response and add them to the `bootstrap` key in the PostHog initialization.
```

> **Tip:** to ensure your request is as quick as possible, use [local evaluation](/docs/feature-flags/local-evaluation) on your server.

## Bootstrapping with a distinct ID

> The `bootstrap` object has two optional arguments: `distinctID` and `isIdentifiedID`.

Reviewer 2: I thought distinctID was not optional 🤔

Author: @neilkakkar ?

> - `distinctID` enables you to set distinct ID of the user during PostHog's initialization. This is useful when you want to ensure the distinct ID on your frontend will be the same as the distinct ID that you called `getAllFlags()` with on your server.

Reviewer 2: ```suggestion
- `distinctID` enables you to set distinct ID of the user during PostHog's initialization. This is useful when you want to ensure the distinct ID on your frontend is the same as the distinct ID that you called `getAllFlags()` with on your server.
```
